File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
/*
# DFS
# 깊이 우선 탐색(Depth-First Search)
- 그래프: 정점(vertex/node)와 간선(edge)의 집합
- 무방향 그래프 / 방향 그래프로 나누어짐
- 인접하다(adjacent): node A에서 간선 하나를 거쳐 node B로 이동할 수 있을 때 A와 B가 인접함
  방향 그래프에서는 한쪽만 성립
- 싸이클(cycle): 간선을 따라 시작한 정점으로 돌아오는 경로
- 가중치 그래프 / 멀티 그래프 등 존재
- 스택 / 재귀함수 이용해 구현 가능
 */
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int [][] house;
    static boolean [][] visit;
    static int cnt, num = 0;
    static int[] dx = {0, -1, 0, 1};
    static int[] dy = {1, 0, -1, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        ArrayList<Integer> list = new ArrayList<>();

        house = new int[N][N];
        visit = new boolean[N][N];
        for (int i = 0; i<N; i++) {
            String s = br.readLine();
            for (int j = 0; j<N; j++) {
                house[i][j] = s.charAt(j) - '0';
            }
        }

        for (int i = 0; i<N; i++) {
            for (int j = 0; j<N; j++) {
                if (house[i][j] == 1 && !visit[i][j]) {
                    dfs(i, j);
                    cnt++;
                    list.add(num);
                    num = 0;
                }
            }
        }
        Collections.sort(list);
        System.out.println(cnt);
        for (int i = 0; i<list.size(); i++) System.out.println(list.get(i));

    }

    static void dfs(int x, int y) {
        num ++;
        visit[x][y] = true;

        for (int i = 0; i<4; i++) {
            int cx = x + dx[i];
            int cy = y + dy[i];

            if (cx >= 0 && cy>= 0 && cx < N && cy < N) {
                if (house[cx][cy] == 1 && !visit[cx][cy]) {
                    dfs(cx, cy);
                }

            }
        }
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque

dx = [0,0,1,-1]
dy = [1,-1,0,0]

def bfs(graph, x, y):
    n = len(graph)
    queue = deque()
    queue.append((x, y))
    graph[x][y] = 0
    cnt = 1
    while queue:
        px, py = queue.popleft()
        for i in range(4):
            nx = px+dx[i]
            ny = py+dy[i]
            if nx<0 or ny<0 or nx>=n or ny>=n:
                continue
            if graph[nx][ny] == 1:
                graph[nx][ny] = 0
                queue.append((nx, ny))
                cnt+=1
    return cnt


N = int(input())
graph = []
for i in range(N):
    graph.append(list(map(int, input())))

cnt =[]
for i in range(N):
    for j in range(N):
        if graph[i][j] == 1:
            cnt.append(bfs(graph, i, j))

cnt.sort()
print(len(cnt))
for i in range(len(cnt)):
    print(cnt[i])